### PR TITLE
[heft] Apply formalism around emitting errors and warnings.

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

File: apps/heft/src/utilities/subprocess/ScopedLoggerManager.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+  resolve: (result: TResult) => void;
+  reject: (error: Error) => void;
+}
+
+export interface IScopedLoggerManagerOptions extends ISubprocessCommunicationManagerBaseOptions {
+  terminalProviderManager: TerminalProviderManager;
+  heftSession?: HeftSession;
+}
+
+export class ScopedLoggerManager extends SubprocessCommunicationManagerBase {
```
Comment (BPapp-MS) [inline apps/heft/src/utilities/subprocess/ScopedLoggerManager.ts:50]: ScopedLoggerManager sounds very similar to the functionality of the LoggingManager class, which also manages ScopedLoggers. Maybe something like SubprocessLoggerManager for this one to communicate that it implements the SubprocessCommunicationManagerBase? Or SubprocessScopedLoggerManager, which is long but accurate, or SubprocessLoggingManager, which follows the same convention as LoggingManager?

File: apps/heft/src/pluginFramework/logging/LoggingManager.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      });
+      this._scopedLoggers.set(loggerName, scopedLogger);
+      return scopedLogger;
+    }
+  }
+
+  public getErrorStrings(): string[] {
+    const result: string[] = [];
+
+    for (const [, scopedLogger] of this._scopedLoggers) {
```
Comment (octogonz) [inline apps/heft/src/pluginFramework/logging/LoggingManager.ts:47]: destructuring is clumsy -- simpler to use `_scopedLoggers.values()`

File: apps/heft/src/utilities/subprocess/ScopedLoggerManager.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+          responseMessage = {
+            type: SCOPED_LOGGER_MANAGER_REQUEST_LOGGER_MESSAGE_TYPE,
+            loggerName: typedMessage.loggerName,
+            error: SubprocessRunnerBase.serializeArg(error) as ISubprocessApiCallArgWithValue<
+              ISerializedErrorValue
+            >
+          };
+        }
+
+        this._sendMessageToSubprocessFn(responseMessage);
```
Comment (octogonz) [inline apps/heft/src/utilities/subprocess/ScopedLoggerManager.ts:136]: `Fn` is not an English word.  Maybe `Callback`?

File: apps/heft/src/utilities/subprocess/SubprocessRunnerBase.ts [hunk 4]
```diff
@@ -157,13 +268,106 @@ export abstract class SubprocessRunnerBase<TSubprocessConfiguration> {
...
+      }
+    }
+
+    throw new Error(
+      'Subprocess communication manager. No communication manager can handle message type ' +
+        `"${message.type}" from subprocess.`
+    );
+  }
+
+  public static serializeArg(arg: unknown): ISubprocessApiCallArg {
```
Comment (octogonz) [inline apps/heft/src/utilities/subprocess/SubprocessRunnerBase.ts:305]: Could you add some unit tests for serializing/deserializing?  Some Jest snapshots of the protocol would be very enlightening.

File: apps/heft/src/pluginFramework/logging/ScopedLogger.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    );
+    this.terminal = new Terminal(this.terminalProvider);
+  }
+
+  /**
+   * {@inheritdoc IScopedLogger.emitError}
+   */
+  public emitError(error: Error): void {
+    this._errors.push(error);
+    this.terminal.writeErrorLine(`ERROR: ${error.message}`);
```
Comment (octogonz) [inline apps/heft/src/pluginFramework/logging/ScopedLogger.ts:84]: I pushed a small change that fixes some lint "errors" to be warnings. More to come in a separate PR.

Repro:
1. Go to `build-tests\heft-node-basic-test\src\index.ts`
2. Add `let x;` to the bottom of that file.
3. Run `rushx build`

You will see this output:
```
[typescript] Encountered 1 TypeScript error:
[typescript] ERROR: src\index.ts:9:5 - (TS6133) 'x' is declared but its value is never read.
[eslint] Encountered 2 ESLint errors:
[eslint] WARNING: src\index.ts:9:5 - (@typescript-eslint/typedef) Expected x to have a type annotation.
[eslint] WARNING: src\index.ts:9:5 - (@typescript-eslint/no-unused-vars) 'x' is defined but never used.
 ---- Compile finished (2828ms) ----
-------------------- Finished (3.781s) --------------------
Encountered 2 warnings:
  [eslint] src\index.ts:9:5 - (@typescript-eslint/typedef) Expected x to have a type annotation.
  [eslint] src\index.ts:9:5 - (@typescript-eslint/no-unused-vars) 'x' is defined but never used.
Encountered 1 errors:
  [typescript] src\index.ts:9:5 - (TS6133) 'x' is declared but its value is never read.
```

Feedback:
- `WARNING` should not be yelling in all caps.  Make it `Warning`
- I'm less sure about `ERROR` but it also seems like yelling. I'm thinking the ideal should be to ensure that the word `Error` appears in logs ONLY when an actual error occurs, thus it doesn't need all caps. I myself have wasted time investigating an "error" log message that is actually expected output in a successful build.  These waste could have been avoided if the toolchain forced a failure whenever that string appears in a log message (e.g. the Jest test that does `console.log("Error: I validated that logging works, yay the test passed!");`.  Other people might have other opinions though.
- `Encountered 2 ESLint errors` should say "warnings" or maybe "issues" if we can't know.
- Why is the compiler classifying `TS6133`  as a hard error? We should investigate that.

Comment (iclanton) [reply apps/heft/src/pluginFramework/logging/ScopedLogger.ts:84]: I want to normalize the way path errors/warnings are handled in a separate PR.

File: apps/heft/src/utilities/subprocess/ScopedLoggerManager.ts (hunk 1 above)
Comment (iclanton) [reply apps/heft/src/utilities/subprocess/ScopedLoggerManager.ts:50]: Let's go with `SubprocessLoggingManager`

**Pull request merged**